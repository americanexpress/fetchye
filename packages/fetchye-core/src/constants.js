/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const ACTION_NAMESPACE = '@fetchye';
export const IS_LOADING = `${ACTION_NAMESPACE}/IS_LOADING`;
export const SET_DATA = `${ACTION_NAMESPACE}/SET_DATA`;
export const DELETE_DATA = `${ACTION_NAMESPACE}/DELETE_DATA`;
export const ERROR = `${ACTION_NAMESPACE}/ERROR`;
export const CLEAR_ERROR = `${ACTION_NAMESPACE}/CLEAR_ERROR`;
