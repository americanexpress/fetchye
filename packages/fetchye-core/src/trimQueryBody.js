const graphQLTerms = ['__args', '__name', '__aliasFor', '__variables', '__directives', '__all_on', '__on', '__typeName', 'mutation'];

export const trimQueryBody = (newQuery, existingQuery) => {
  const trimmedQuery = {};
  Object.keys(newQuery).forEach((key) => {
    if (newQuery[key] && !existingQuery[key]) {
      trimmedQuery[key] = newQuery[key];
    } else if (graphQLTerms.indexOf(key) > -1) {
      trimmedQuery[key] = newQuery[key];
    } else if (typeof newQuery[key] === 'object' && !Array.isArray(newQuery[key]) && newQuery[key] !== null) {
      const cutQueryObject = trimQueryBody(newQuery[key], existingQuery[key]);
      const {
        __args, __name, ...rest
      } = cutQueryObject;
      if (Object.keys(rest).length > 0) {
        trimmedQuery[key] = cutQueryObject;
      }
    }
  });
  return trimmedQuery;
};
