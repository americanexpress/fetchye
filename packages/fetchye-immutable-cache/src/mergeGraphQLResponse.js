import merge from 'lodash.merge';

export const mergeGraphQLResponses = (existingResponse, newResponse) => {
  if (!existingResponse) return newResponse;

  return {
    ...existingResponse,
    ...newResponse,
    body: merge(existingResponse.body, newResponse.body),
  };
};
