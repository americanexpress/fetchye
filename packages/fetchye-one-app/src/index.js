export { makeOneServerFetchye } from './makeOneServerFetchye';
export OneCache, { oneCacheSelector } from './OneCache';
export OneFetchyeProvider from './OneFetchyeProvider';
