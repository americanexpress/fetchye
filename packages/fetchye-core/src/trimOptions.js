import { jsonToGraphQLQuery } from 'json-to-graphql-query';
import { trimQueryBody } from './trimQueryBody';

export const trimOptions = (options) => {
  const { existingQuery, isGraphQL } = options || {};
  const {
    body: requestBody,
    ...restOfOptions
  } = options || {};
  const {
    query,
    ...restOfBody
  } = requestBody || {};

  if (isGraphQL && query && existingQuery) {
    return {
      ...restOfOptions,
      body: JSON.stringify({
        query: jsonToGraphQLQuery(trimQueryBody(query, existingQuery), { pretty: true }),
        ...restOfBody,
      }),
    };
  }
  if (isGraphQL && query) {
    return {
      ...restOfOptions,
      body: JSON.stringify({
        query: jsonToGraphQLQuery(query, { pretty: true }),
        ...restOfBody,
      }),
    };
  }
  return options;
};
