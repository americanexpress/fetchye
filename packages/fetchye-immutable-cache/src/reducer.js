/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { Map as iMap, Set as iSet } from 'immutable';
import {
  IS_LOADING,
  SET_DATA,
  DELETE_DATA,
  ERROR,
  UPDATE_DATA,
  SET_QUERY,
  CLEAR_ERROR,
  ACTION_NAMESPACE,
} from 'fetchye-core';
import { mergeGraphQLResponses } from './mergeGraphQLResponse';

// eslint-disable-next-line default-param-last -- the first default param value takes care of explicitly calling this function with `undefined` the second param can't be defaulted as it must be provided
export function fetchyeReducer(state = iMap({
  errors: iMap(),
  loading: iSet(),
  data: iMap(),
  query: iMap(),
}), action) {
  if (!action.type.startsWith(ACTION_NAMESPACE)) {
    return state;
  }
  switch (action.type) {
    case DELETE_DATA:
      return state.deleteIn(['data', action.hash]);
    case CLEAR_ERROR:
      return state.deleteIn(['errors', action.hash]);
    case ERROR:
      return state.withMutations(
        (nextState) => nextState
          .setIn(['errors', action.hash], action.error)
          .deleteIn(['data', action.hash])
          .deleteIn(['loading', action.hash])
      );
    case IS_LOADING:
      return state.updateIn(['loading'], (loading) => loading.add(action.hash));
    case SET_DATA:
      return state.withMutations(
        (nextState) => nextState
          .setIn(['data', action.hash], action.value)
          .deleteIn(['loading', action.hash])
          .deleteIn(['errors', action.hash])
      );
    case UPDATE_DATA:
      return state.withMutations(
        (nextState) => nextState
          .updateIn(['data', action.hash], (existingData) => mergeGraphQLResponses(existingData, action.value))
          .deleteIn(['loading', action.hash])
          .deleteIn(['errors', action.hash])
      );
    case SET_QUERY:
      return state.setIn(['query', action.hash], action.query);
    default:
      return state;
  }
}

export default fetchyeReducer;
