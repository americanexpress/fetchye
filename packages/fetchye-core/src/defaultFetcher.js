/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { trimOptions } from './trimOptions';

// Object.fromEntries is not compatible with Node v10
// provide our own lightweight solution
export const headersToObject = (headers = []) => [...headers]
  .reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {});
export const defaultFetcher = async (fetchClient, key, initialOptions) => {
  let res;
  let payload;
  let error;
  const options = trimOptions(initialOptions);
  try {
    res = await fetchClient(key, options);
    let body = await res.text();
    try {
      body = JSON.parse(body);
    } catch (e) {
      // body will still be the text from the response, so no action needed here
    }
    payload = {
      body,
      ok: res.ok,
      headers: headersToObject(res.headers),
      status: res.status,
    };
  } catch (requestError) {
    // eslint-disable-next-line no-console -- error useful to developer in specific error case
    console.error(requestError);
    error = requestError;
  }
  return {
    payload,
    error,
  };
};

export default defaultFetcher;
