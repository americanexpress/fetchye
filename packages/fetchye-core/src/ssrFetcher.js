import { defaultFetcher } from './defaultFetcher';

export const ssrFetcher = async (...params) => {
  const { payload } = await defaultFetcher(...params);
  return {
    payload,
    error: null,
  };
};

export default ssrFetcher;
