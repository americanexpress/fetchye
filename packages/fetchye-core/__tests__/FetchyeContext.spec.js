/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import createSharedReactContext from 'create-shared-react-context';
import { SHARED_CONTEXT_ID } from '../src/FetchyeContext';

jest.mock('create-shared-react-context', () => jest.fn());

describe('FetchyeContext', () => {
  it('is called with SHARED_CONTEXT_ID', () => {
    expect(createSharedReactContext).toHaveBeenCalledWith(null, SHARED_CONTEXT_ID);
    expect(createSharedReactContext).toMatchSnapshot();
  });
});
