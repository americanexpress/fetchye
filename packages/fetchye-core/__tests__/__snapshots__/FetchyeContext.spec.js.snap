// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchyeContext is called with SHARED_CONTEXT_ID 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      null,
      "FetchyeContext",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
